### DISCUSSION TITLE
Running Avalonia inside FNA/MonoGame

### DISCUSSION THREAD

**Discussion by jordi-z80** [AvaloniaUI/Avalonia#16124] [Q&A]
I've been reading about the way Avalonia works and I'm wondering if it'd be possible to run Avalonia inside FNA or MonoGame. I think it'd be far more useful that the other way around, which is what people usually tries to. My idea would be rendering the Avalonia UI into a texture with alpha and render them on top of the .

Would this be possible?

Also, is there documentation on how to create a backend like this?

Thanks!

**Comment by stevemonaco** [comment]
Check out https://github.com/MrJul/Estragonia for an example of this.

It can require quite a bit of work because you also need to integrate pointer (mouse), keyboard, and several other things. IMO, it's expected that any implementer has fairly deep knowledge of the current Avalonia as many APIs you'll need are marked `[PrivateApi]` or `[Unstable]`, so docs would never keep up.

**Comment by jordi-z80** [comment]
Hi Steve, thanks for your answer.

I've been checking the example and tried to reproduce it in a KNI environment (a fork of MonoGame). I've managed to get a UserControl loading and the layout is supposedly being "laid out" (according to logs), but I'm having some big problems with the rendering (it's like both openGL environments are fighting each other).

However in the last hours I've been playing with the headless version of Avalonia, and it seems a lot easier and promising - probably at the expense of less flexibility / performance. 
I've been able to easily use it in conjunction with the mouse/keyboard functions to have something functional (although very slow because of the gpu->cpu->gpu memory of the skia image).

My doubt now is if Avalonia Headless is fit for such a task, and (when used in conjunction with UseSkia) if it's hardware accelerated.
If it is, it'll be just a matter of being able to use the skia surface directly as a texture for speed reasons and find a way to "simulate time pass" in the UI (some color animations are not being updated right now, I may need to call something else)

**Comment by jordi-z80** [comment]
According to this https://github.com/AvaloniaUI/Avalonia/discussions/15761  headless is not hardware accelerated, so back to square one :(

**Reply by stevemonaco** [reply]
AFAIK, Headless is intended for integration testing of Avalonia applications. Avalonia.LinuxFramebuffer exists, but I think that's mostly intended for presenting images to embedded devices, not meant for game engines. That's why I suggested creating a new Platform backend as that will give you the greatest ability to integrate though I'm unfamiliar with how much CPU<->GPU copying will be necessary.
